### PR TITLE
mem: Atomic ops to same address

### PR DESCRIPTION
**Pull request by BujSet**
Augmenting the DataBlock class with a change log structure to record the effects of atomic operations on a data block and service these changes if the atomic operations require return values.

Although the operations are atomic, the coalescer need not send unique memory requests for each operation. Atomic operations within a wavefront to the same address are now coalesced into a single memory request. The response of this request carries all the necessary information to provide the requesting lanes unique values as a result of their individual atomic operations. This helps reduce contention for request and response queues in simulation.

Previously, only the final value of the datablock after all atomic ops to the same address was visible to the requesting waves. This change corrects this behavior by allowing each wave to see the effect of this individual atomic op is a return value is necessary.

### CODE REVIEW

**Comment by powerjg** [comment]
I know it's different, but I think that #177 is quite related. Let's make sure there are no conflicts between these changes.

**Comment by BujSet** [comment]
Forgot to retain the changeId in the commit message. Added it back.

**Comment by BujSet** [comment]
> I know it's different, but I think that #177 is quite related. Let's make sure there are no conflicts between these changes.

Looking through the files changed, I don't think the there should be any merge-conflict issues. I haven't tested this change with the CHI protocol, but I suspect CHI has the same issue as the VIPER protocol (where waves that request atomic updates to the same address get the same response instead of a unique responses after each atomic update). Will test this out after [#177](https://github.com/gem5/gem5/pull/177/) gets merged into develop (to prevent conflicts now).

**Comment by mattsinc** [comment]
@BujSet I am away for the next few days, and while I reviewed it at AMD previously, I will review here again after Monday.

File: src/mem/ruby/common/DataBlock.cc
Comment (mattsinc) [inline src/mem/ruby/common/DataBlock.cc:89]: minor nit: please add parentheses around these if statements.  It will bite us in the butt later if we don't.

Technically not required though.

File: src/mem/ruby/system/GPUCoalescer.cc
Comment (mattsinc) [inline src/mem/ruby/system/GPUCoalescer.cc]: maybe not part of this commit, but AD --> ADD?

File: src/mem/ruby/system/GPUCoalescer.cc
Comment (mattsinc) [inline src/mem/ruby/system/GPUCoalescer.cc]: is there a case where the type is an atomic return but isAtomicOp returns false?

**Review by mattsinc — CHANGES_REQUESTED** [review]
I don't see anything overlapping relative to the other patch, but I do have some small changes that would be helpful here.

File: src/mem/ruby/common/DataBlock.cc
Comment (giactra) [inline src/mem/ruby/common/DataBlock.cc]: This is a duplicate

File: src/mem/ruby/common/DataBlock.cc
Comment (giactra) [inline src/mem/ruby/common/DataBlock.cc]: This is probably not needed as we would just skip the for loop

File: src/mem/ruby/common/DataBlock.hh
Comment (giactra) [inline src/mem/ruby/common/DataBlock.hh]: Ditto

File: src/mem/ruby/common/DataBlock.hh
Comment (giactra) [inline src/mem/ruby/common/DataBlock.hh:109]: Wouldn't it make sense to use a unique_ptr instead?

File: src/mem/ruby/common/WriteMask.hh
Comment (giactra) [inline src/mem/ruby/common/WriteMask.hh]: This seems a GPU specific comment

File: src/mem/ruby/common/WriteMask.cc
Comment (giactra) [inline src/mem/ruby/common/WriteMask.cc:72]: I am wondering whether you could push most of the bookeping to the AtomicOpFunctor....

Comment (BujSet) [reply src/mem/ruby/common/WriteMask.cc:72]: I think it could be done. To do this though, the response messages would have to carry the WriteMask back up the hierarchy, since the vector of functors are held by the WriteMask itself. Since the DataBlock is already a part of the response message, I opted to put the logging structure there. Would it make more sense to use the AtomicOpFunctors? I think putting it there might make it a little clunky: in the GPUCoalescer, we would read from the return WriteMasks atomicVector of functors instead of the from the datablock.

File: src/mem/ruby/system/GPUCoalescer.cc
Comment (BujSet) [reply src/mem/ruby/system/GPUCoalescer.cc]: I believe that that case should never happen, but I vaguely remember seeing this pop up on some of the tester benchmarks, but I'm failing to remember which ones. Will test this out and see if the if-check is necessary.

Comment (BujSet) [reply src/mem/ruby/system/GPUCoalescer.cc]: Looks like I remembered wrong, and I did not see a case when the packet->isAtomicOp() needs after already seeing that the ruby request type is one of the atomic types. Put in some assert statements so that this assumption is clear moving forward.

**Comment by BujSet** [comment]
@abmerop Tagging you here as well.

**Review by abmerop — COMMENTED** [review]
This looks fine to me, but of course I have seen multiple iterations of this code.

File: src/mem/ruby/common/DataBlock.hh
Comment (abmerop) [reply src/mem/ruby/common/DataBlock.hh:109]: The DataBlock can be copied so doesn't this just make things more complex?

**Comment by BobbyRBruce** [comment]
@mattsinc : Is this ok with you? You had some suggestions which I believe have been incorporated. If so I'll merge this.

**Comment by mattsinc** [comment]
I was waiting for Giacomo to chime in since Giacomo was the one requesting
the changes.

**Comment by mattsinc** [comment]
> 

@giactra just checking if you are ok with the changes @BujSet made?

File: src/mem/ruby/common/WriteMask.cc
Comment (giactra) [reply src/mem/ruby/common/WriteMask.cc:72]: > I think it could be done. To do this though, the response messages would have to carry the WriteMask back up the hierarchy, since the vector of functors are held by the WriteMask itself.
Since the DataBlock is already a part of the response message, I opted to put the logging structure there. Would it make more sense to use the AtomicOpFunctors? I think putting it there might make it a little clunky: in the GPUCoalescer, we would read from the return WriteMasks atomicVector of functors instead of the from the datablock.

Apologies for the late reply, I was on a small break from work. Yeah I understand, and binding the log of changes to the block itself makes a lot of sense. My original message was just meant to move the three lines to the "prologue" of the functor to keep the WriteMask cleaner

        block_update = new uint8_t[mSize];
        std::memcpy(block_update, p, mSize);
        log.push_back(block_update);

But I realise it is a little save...

**Review by giactra — APPROVED** [review]
LGTM

**Review by mattsinc — APPROVED** [review]
Thanks @giactra !  @BujSet you will need to rebase before you can merge this in, FYI.

**Comment by BobbyRBruce** [comment]
> Thanks @giactra ! @BujSet you will need to rebase before you can merge this in, FYI.

Do they? It's clean merge from what I can see.

**Comment by mattsinc** [comment]
> > Thanks @giactra ! @BujSet you will need to rebase before you can merge this in, FYI.
> 
> Do they? It's clean merge from what I can see.

Interesting, it is clean now but when I sent that, I saw that rebasing to the head of develop was required ...

**Comment by BobbyRBruce** [comment]
> > > Thanks @giactra ! @BujSet you will need to rebase before you can merge this in, FYI.
> > 
> > 
> > Do they? It's clean merge from what I can see.
> 
> Interesting, it is clean now but when I sent that, I saw that rebasing to the head of develop was required ...

Huh, I'll look out for that. No worries, I just thought there was some reason it needed rebasing.

Since this now has 2 approvals, and I don't see any outstanding comments to address, I'll merge it,

Thank you for all great work guys!

**Pull request merged**
